### PR TITLE
[SPARK-23005][Core] Improve RDD.take on small number of partitions

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?
In current implementation of RDD.take, we overestimate the number of partitions we need to try by 50%:
`(1.5 * num * partsScanned / buf.size).toInt`
However, when the number is small, the result of `.toInt` is not what we want. 
E.g, 2.9 will become 2, which should be 3.
Use Math.ceil to fix the problem.

Also clean up the code in RDD.scala.

## How was this patch tested?

Unit test

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala [hunk 1]
```diff
@@ -985,7 +985,7 @@ abstract class RDD[T: ClassTag](
   def subtract(
       other: RDD[T],
       p: Partitioner)(implicit ord: Ordering[T] = null): RDD[T] = withScope {
-    if (partitioner == Some(p)) {
+    if (partitioner.contains(p)) {
```
Comment (hvanhovell) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: Do we still support scala 2.10? If we do, this will fail compilation.

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/rdd/RDD.scala]: Actually I think the previous code is more readable.

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala [hunk 2]
```diff
@@ -1345,13 +1346,12 @@ abstract class RDD[T: ClassTag](
           if (buf.isEmpty) {
             numPartsToTry = partsScanned * scaleUpFactor
           } else {
-            // the left side of max is >=1 whenever partsScanned >= 2
-            numPartsToTry = Math.max((1.5 * num * partsScanned / buf.size).toInt - partsScanned, 1)
+            // As left > 0, numPartsToTry is always >= 1
```
Comment (hvanhovell) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: This is the same a s SparkPlan. executeTake(). Should we also fix that?

**Review by srowen — APPROVED** [review]

**Comment by cloud-fan** [comment]
LGTM, merging to master/2.3!

**Pull request closed without merging**